stats/opencensus/e2e_test.go

```diff
@@ -0,0 +1,773 @@
+	}
+	return true
+}
+
+// compareRows compares rows with respect to the information desired to test.
+// Both the tags representing the rows and also the data of the row are tested
+// for equality. Rows are in nondeterministic order when ExportView is called,
+// but handled because ExportView is called every default metrics reporting
+// interval, which will work eventually in a reasonable manner (i.e. doesn't
+// flake).
+func compareRows(rows []*view.Row, rows2 []*view.Row) bool {
+	if rows == nil && rows2 == nil {
+		return true
+	}
+	if (rows != nil) != (rows2 != nil) {
+		return false
+	}
+	if len(rows) != len(rows2) {
+		return false
+	}
+
+	for i, row := range rows {
+		if !cmp.Equal(row.Tags, rows2[i].Tags, cmp.Comparer(func(a tag.Key, b tag.Key) bool {
+			return a.Name() == b.Name()
+		})) {
+			return false
+		}
+		if !compareData(row.Data, rows2[i].Data) {
+			return false
+		}
+	}
+	return true
+}
+
+// compareData returns whether the two aggregation data's are equal to each
+// other with respect to parts of the data desired for correct emission. The
+// function first makes sure the two types of aggregation data are the same, and
+// then checks the equality for the respective aggregation data type.
+func compareData(ad view.AggregationData, ad2 view.AggregationData) bool {
+	if ad == nil && ad2 == nil {
+		return true
+	}
+	if (ad != nil) != (ad2 != nil) {
+		return false
+	}
+	switch ad1 := ad.(type) {
+	case *view.DistributionData:
+		dd2, ok := ad2.(*view.DistributionData)
+		if !ok {
+			return false
+		}
+		// Count and Count Per Buckets are reasonable for correctness,
+		// especially since we verify equality of bucket endpoints elsewhere.
+		if ad1.Count != dd2.Count {
+			return false
+		}
+		for i, count := range ad1.CountPerBucket {
+			if count != dd2.CountPerBucket[i] {
+				return false
+			}
+		}
+	case *view.CountData:
+		cd2, ok := ad2.(*view.CountData)
+		if !ok {
+			return false
+		}
+		return ad1.Value == cd2.Value
+
+	// gRPC open census plugin does not have these next two types of aggregation
+	// data types present, for now just check for type equality between the two
+	// aggregation data points.
+	case *view.SumData:
+		_, ok := ad2.(*view.SumData)
+		if !ok {
+			return false
+		}
+	case *view.LastValueData:
+		_, ok := ad2.(*view.LastValueData)
+		if !ok {
+			return false
+		}
+	}
+	return true
+}
+
+func (vi *viewInformation) Equal(vi2 *viewInformation) bool {
+	return vi.equal(vi2)
+}
+
+// seenViewPresent checks if the seen views contain the desired information for
+// the view within the scope of a certain time. Returns an nil slice if correct
+// information found, non nil slice of at least length 1 if correct information
+// not found within a timeout.
+func seenViewMatches(fe *fakeExporter, viewName string, wantVI *viewInformation) []error {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	var errs []error
+	for ctx.Err() == nil {
+		errs = nil
```

Comment:
`errs = errs[:0]` would keep the backing memory and make the `appends` cheaper.

But this function confuses me.

Why are you using a slice?  You only have one possible `append` call each time through the loop, and it's cleared at the top of the loop.